File: FilmesAPI.FSharp/Controllers/FilmeController.fs
namespace FilmesAPI.FSharp.Controllers

open Microsoft.AspNetCore.Mvc
open FilmesAPI.FSharp.Models
open Microsoft.Extensions.Logging
open FilmesAPI.FSharp
open FilmesAPI.FSharp.Data
open AutoMapper
open FilmesAPI.FSharp.Data.Dtos
open System

[<ApiController>]
[<Route("[controller]")>]
type FilmeController(dbContext : FilmeContext, mapper: IMapper) =
    inherit ControllerBase()

    [<HttpPost>]
    member this.AdicionaFilme ([<FromBody>]filmeDto: CreateFilmeDto) =
        let filme = mapper.Map<Filme>(filmeDto)
        dbContext.Filmes.Add filme |> ignore
        dbContext.SaveChanges() |> ignore
        this.CreatedAtAction("RecuperaFilmePorId", {| id = filme.Id |}, filme)

    [<HttpGet>]
    member this.RecuperaFilmes () = this.Ok(dbContext.Filmes)

    [<HttpGet("{id}", Name = "RecuperaFilmePorId")>]
    member this.RecuperaFilmePorId (id: int) : IActionResult = 
        match dbContext.Filmes |> Seq.tryFind (fun f -> f.Id = id) with
        | Some filme -> 
            let filmeDto = filme |> mapper.Map<ReadFilmeDto>
            this.Ok({filmeDto with HoraDaConsulta = DateTime.Now})
        | None -> this.NotFound()

    [<HttpPut("{id}")>]
    member this.AtualizaFilme (id: int) ([<FromBody>]filmeDto: UpdateFilmeDto) : IActionResult =
        try
            let filme = dbContext.Filmes.Find(id)
            let filmeNovo = mapper.Map(filmeDto, filme)
            dbContext.Entry(filme).CurrentValues.SetValues({filmeNovo with Id = filme.Id})
            dbContext.SaveChanges() |> ignore
            this.NoContent()
        with
        | :? System.ArgumentNullException -> this.NotFound()

    [<HttpDelete("{id}")>]
    member this.DeletaFilme id : IActionResult =
        match dbContext.Filmes |> Seq.tryFind (fun f -> f.Id = id) with
        | Some filme -> 
            dbContext.Remove(filme) |> ignore
            dbContext.SaveChanges() |> ignore
            this.NoContent()
        | None -> this.NotFound()

File: FilmesAPI/Controllers/FilmeController.cs
using AutoMapper;
using FilmesAPI.Data;
using FilmesAPI.Data.Dtos.Filme;
using FilmesAPI.Models;
using FilmesAPI.Services;
using Microsoft.AspNetCore.Mvc;

namespace FilmesAPI.Controllers;

[ApiController]
[Route("[controller]")]
public class FilmeController : ControllerBase
{
    private readonly FilmeService _filmeService;

    public FilmeController(FilmeService filmeService)
    {
        _filmeService = filmeService;
    }

    [HttpPost]
    public IActionResult AdicionaFilme([FromBody] CreateFilmeDto filmeDto)
    {
        var readDto = _filmeService.AdicionaFilme(filmeDto);
        return CreatedAtAction(nameof(RecuperaFilmePorId), new { Id = readDto.Id }, readDto);
    }

    [HttpGet]
    public IActionResult RecuperaFilmes([FromQuery] int? classificacaoEtaria = null)
    {
        var readDto = _filmeService.RecuperaFilmes(classificacaoEtaria);
        if (readDto is not null) return Ok(readDto);
        return NotFound();
    }

    [HttpGet("{id}")]
    public IActionResult RecuperaFilmePorId(int id)
    {
        var readDto = _filmeService.RecuperaFilmePorId(id);
        if (readDto is not null) return Ok(readDto);
        return NotFound();
    }

    [HttpPut("{id}")]
    public IActionResult AtualizaFilme(int id, [FromBody] UpdateFilmeDto filmeDto)
    {
        var resultado = _filmeService.AtualizaFilme(id, filmeDto);
        if (resultado.IsFailed) return NotFound();

        return NoContent();
    }

    [HttpDelete("{id}")]
    public IActionResult DeletaFilme(int id)
    {
        var resultado = _filmeService.DeletaFilme(id);
        if (resultado.IsFailed) return NotFound();

        return NoContent();
    }
}

